types/src/program.ts

```diff
@@ -1,35 +1,49 @@
 import path from "path";
 import ts from "typescript";
 
-export function createMemoryProgram(source: string): [ts.Program, string] {
+interface MemorySourceFile {
+  source: string;
+  sourceFile: ts.SourceFile;
+}
+
+// Creates a TypeScript program form in-memory source files. Accepts a Map of
+// fully-resolved "virtual" paths to source code.
+export function createMemoryProgram(sources: Map<string, string>): ts.Program {
```

Comment:
This is very cool, but is it definitely necessary? What's the problem with writing to a temporary file?